File: bearAndUpDown.py
def find_bad_spots(s):
    bad_spots = []
    for i in range(len(s)-1):
        if i % 2 == 0:
            if s[i] >= s[i+1]:
                # odd
                bad_spots.append(i)
        else:
            if s[i] <= s[i+1]:
                # even
                bad_spots.append(i)
    return bad_spots

def create_swap_points(bad_spots, end):
    swap_points = []
    for i in bad_spots:
        if i == 0:
            swap_points.append(i)
            swap_points.append(i+1)

        elif i == end:
            swap_points.append(i-1)
            swap_points.append(i)

        else:
            swap_points.append(i-1)
            swap_points.append(i)
            swap_points.append(i+1)

    return list(set(swap_points))

def check_if_bad(i, j, s, bad_spots):
    points = list(set([i-1, i, i+1, j-1, j, j+1]))
    # print "points:", points
    # print "s:", s, i, j

    for p in bad_spots:
        if p not in points:
            return False

    for p in points:
        # print p
        if p == len(s)-1:
            p -= 1

        if p >= 0 and p < len(s)-1:
            if p % 2 == 0:
                if s[p] >= s[p+1]:
                    # odd
                    return False
            else:
                if s[p] <= s[p+1]:
                    # even
                   return False
    return True


def main():
    n = int(raw_input())
    t = [int(i) for i in raw_input().split()]
    count = 0

    swaps = set()
    bad_spots = find_bad_spots(t)

    if len(bad_spots) > 4:
        pass

    elif len(bad_spots) == 4:
        swap_points = create_swap_points(bad_spots, len(t)-1)
        for i in range(len(swap_points)-1):
            for j in range(i+1, len(swap_points)):
                idx_a, idx_b = ((swap_points[i], swap_points[j]) if swap_points[i] < swap_points[j] else (swap_points[j], swap_points[i]))
                tmp = t[idx_a]
                t[idx_a] = t[idx_b]
                t[idx_b] = tmp

                if check_if_bad(idx_a, idx_b, t, bad_spots):
                    swaps.add((idx_a, idx_b))
                    count += 1

                tmp = t[idx_a]
                t[idx_a] = t[idx_b]
                t[idx_b] = tmp

    else:
        swap_points = create_swap_points(bad_spots, len(t)-1)
        for i in range(len(swap_points)):
            for j in range(len(t)):
                idx_a, idx_b = ((swap_points[i], j) if swap_points[i] < j else (j, swap_points[i]))
                if idx_a != idx_b:
                    tmp = t[idx_a]
                    t[idx_a] = t[idx_b]
                    t[idx_b] = tmp

                    if check_if_bad(idx_a, idx_b, t, bad_spots):
                        # print "a:", idx_a, "b:", idx_b
                        swaps.add((idx_a, idx_b))
                        count += 1

                    tmp = t[idx_a]
                    t[idx_a] = t[idx_b]
                    t[idx_b] = tmp

    # print swaps
    print len(swaps)
    return

main()

File: bearAndUpDown.cpp
#include <vector>
#include <list>
#include <map>
#include <set>
#include <queue>
#include <deque>
#include <stack>
#include <bitset>
#include <algorithm>
#include <functional>
#include <numeric>
#include <utility>
#include <sstream>
#include <iostream>
#include <iomanip>
#include <cstdio>
#include <cmath>
#include <cstdlib>
#include <cstring>
#include <ctime>

using namespace std;

bool is_bad_spot(long long int i, vector<long long int> s){
	// cout<<i<<" "<<s[i]<<" "<<s[i+1]<<endl;
	if(i%2 == 0){
		if(s[i] >= s[i+1]){
			// cout<<"trye odd"<<endl;
			return true;
		}
	}
	else{
		if(s[i] <= s[i+1]){
			// cout<<"true even"<<endl;
			return true;
		}
	}
	// cout<<"s"<<endl;
	// i++;
	return false;
}

bool has_bad_spots(long long int idx_a, long long int idx_b, vector<long long int> v, vector<long long int>bad_spots){
	long long int i,j,p;
	vector<long long int> points;

	points.push_back(idx_a-1);
	points.push_back(idx_a);
	points.push_back(idx_a+1);
	points.push_back(idx_b-1);
	points.push_back(idx_b);
	points.push_back(idx_b+1);

	for(i=0;i<bad_spots.size();i++){
		if(find(points.begin(), points.end(), bad_spots[i])==points.end()){
			return true;
		}
	}

	for(i=0;i<points.size();i++){
		p = points[i];
		if(p == v.size()-1){
			p--;
		}

		if(p >= 0 && p < v.size()-1){
			if(is_bad_spot(p, v)){
				return true;
			}
		}
	}

	return false;
}

vector<long long int> find_bad_spots(vector<long long int>v){
	long long int i,j,k;
	vector<long long int> bad_spots;

	for(i=0;i<=v.size()-2;i++){
		// cout<<"i:"<<i<<" "<<v[i]<<" "<<v[i+1]<<endl;
		if(is_bad_spot(i, v)){
			// cout<<"pushing"<<endl;
			bad_spots.push_back(i);
		}
		// cout<<"pushed "<<v.size()<<endl;
	}
	// cout<<"return";
	return bad_spots;

}

set<long long int> create_swap_spots(vector<long long int> bad_spots, long long int sz){
	long long int i, idx;
	set<long long int> swap_spots;

	for(i=0;i<bad_spots.size();i++){
		idx = bad_spots[i];
		if(idx==0){
			swap_spots.insert(idx);
			swap_spots.insert(idx+1);
		}

		else if(idx==sz){
			swap_spots.insert(idx-1);
			swap_spots.insert(idx);
		}

		else{
			swap_spots.insert(idx-1);
			swap_spots.insert(idx);
			swap_spots.insert(idx+1);
		}
	}

	return swap_spots;
}

typedef struct
{
	long long int x;
	long long int y;
}point;

bool contains(vector<point> v, point a){
	long long int i;
	point b;
	for(i=0;i<v.size();i++){
		b = v[i];
		if(a.x == b.x and a.y == b.y){
			return true;
		}
	}
	return false;
}

int main(){
	long long int i,j,k,n,count=0,idx_a, idx_b, tmp;
	vector<long long int>v, bad_spots;
	vector<point> swaps;
	set<long long int> swap_spots;
	set<long long int>::iterator it_a, it_b;
	cin>>n;

	for(i=0;i<n;i++){
		cin>>j;
		v.push_back(j);
	}

	bad_spots = find_bad_spots(v);

	// for(i=0;i<bad_spots.size();i++){
	// 	cout<<bad_spots[i]<<" ";
	// }
	// cout<<endl;

	// cout<<bad_spots.size()<<endl;

	if(bad_spots.size() > 4){
		cout<<0<<endl;
		return 0;
	}

	else if(bad_spots.size() == 4){
		// cout<<"else if"<<endl;
		swap_spots = create_swap_spots(bad_spots, v.size()-1);
		// cout<<"created swap spots"<<endl;
		for(it_a=swap_spots.begin();it_a!=--swap_spots.end();it_a++){
			for(it_b=++it_a;it_b!=swap_spots.end();it_b++){
				idx_a = *it_a;
				idx_b = *it_b;

				tmp = v[idx_a];
                v[idx_a] = v[idx_b];
                v[idx_b] = tmp;

                if(!has_bad_spots(idx_a, idx_b, v, bad_spots)){
                	count++;
                }

                tmp = v[idx_a];
                v[idx_a] = v[idx_b];
                v[idx_b] = tmp;
			}
		}
	}

	else{
		swap_spots = create_swap_spots(bad_spots, v.size()-1);
		for(it_a=swap_spots.begin();it_a!=swap_spots.end();it_a++){
			for(i=0;i<v.size();i++){
				idx_a = *it_a;
				idx_b = i;

				if(idx_a>idx_b){
					tmp = idx_a;
					idx_a = idx_b;
					idx_b = tmp;
				}

				if(idx_a == idx_b)
					continue;

				tmp = v[idx_a];
                v[idx_a] = v[idx_b];
                v[idx_b] = tmp;

                if(!has_bad_spots(idx_a, idx_b, v, bad_spots)){
                	point sp;
                	sp.x = idx_a;
                	sp.y = idx_b;
                	if(!contains(swaps, sp)){
                		count++;
                		swaps.push_back(sp);	
                	}
                }

                tmp = v[idx_a];
                v[idx_a] = v[idx_b];
                v[idx_b] = tmp;
			}
		}
	}

	// for(it_a=swap_spots.begin();it_a!=swap_spots.end();it_a++){
	// 	cout<<*it_a<<" ";
	// }
	// cout<<endl;

	cout<<count<<endl;
	return 0;
}